## Data Wrangling

In [356]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import tweepy
import requests

#### Importing archive File

In [357]:
dogs_df = pd.read_csv('twitter-archive-enhanced.csv')
dogs_df.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193 NaN                    NaN                    
1  892177421306343426 NaN                    NaN                    
2  891815181378084864 NaN                    NaN                    
3  891689557279858688 NaN                    NaN                    
4  891327558926688256 NaN                    NaN                    

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                                                               source  \
0  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
3  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
4  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                 text  \
0  This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU                                                        
1  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
2  This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB                    
3  This is Darla. She commenced a snooze mid meal. 13/10 happens to the best of us https://t.co/tD36da7qLQ                                                              
4  This is Franklin. He would like you to stop calling him "cute." He is a very fierce shark and should be respected as such. 12/10 #BarkWeek https://t.co/AtUZn91f7f   

   retweeted_status_id  retweeted_status_user_id retweeted_status_timestamp  \
0 NaN                  NaN                        NaN                         
1 NaN                  NaN                        NaN                         
2 NaN                  NaN                        NaN                         
3 NaN                  NaN                        NaN                         
4 NaN                  NaN                        NaN                         

                                                                                                                     expanded_urls  \
0  https://twitter.com/dog_rates/status/892420643555336193/photo/1                                                                   
1  https://twitter.com/dog_rates/status/892177421306343426/photo/1                                                                   
2  https://twitter.com/dog_rates/status/891815181378084864/photo/1                                                                   
3  https://twitter.com/dog_rates/status/891689557279858688/photo/1                                                                   
4  https://twitter.com/dog_rates/status/891327558926688256/photo/1,https://twitter.com/dog_rates/status/891327558926688256/photo/1   

   rating_numerator  rating_denominator      name doggo floofer pupper puppo  
0  13                10                  Phineas   None  None    None   None  
1  13                10                  Tilly     None  None    None   None  
2  12                10                  Archie    None  None    None   None  
3  13                10                  Darla     None  None    None   None  
4  12                10                  Franklin  None  None    None   None

In [358]:
dogs_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [359]:
dogs_df.tweet_id.nunique()

2356

#### Wrangling tweet image prediction image_predictions.tsv file from Udacity URL

In [360]:
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
r = requests.get(url)

with open('image_predictions.tsv', 'wb') as file:
    file.write(r.content)

In [361]:
predict_df = pd.read_csv('image_predictions.tsv', sep='\t')
predict_df.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0  1        Welsh_springer_spaniel  0.465074  True    collie               
1  1        redbone                 0.506826  True    miniature_pinscher   
2  1        German_shepherd         0.596461  True    malinois             
3  1        Rhodesian_ridgeback     0.408143  True    redbone              
4  1        miniature_pinscher      0.560311  True    Rottweiler           

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665  True    Shetland_sheepdog    0.061428  True    
1  0.074192  True    Rhodesian_ridgeback  0.072010  True    
2  0.138584  True    bloodhound           0.116197  True    
3  0.360687  True    miniature_pinscher   0.222752  True    
4  0.243682  True    Doberman             0.154629  True

#### Wrangling twitter retweets and likes using twiter API

In [7]:
consumer_key = '*************************'
consumer_secret = '**************************************************'

access_token = '**************************************************'
access_secret = '*********************************************'

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit= True, wait_on_rate_limit_notify  = True)

In [8]:
from pprint import pprint
import json
import os
import time

def dump_tweet_json(tweet_id, file):
    """
    query the tweet json file specified by twitter id 'tweet_id'
    using twitter API tweepy, parse the json object and write it to the
    file 'file'
    
    Warning! file is assumed open in writing mode
    """
    tweet_status = api.get_status(tweet_id, tweet_mode='extended')
    json.dump(tweet_status._json, file)
    

In [9]:
deleted_tweets_ids = []
rate_limit_exceeded_tweets  = []

with open('tweet_json.txt', 'w') as file:
    
    ## First Path of Querying tweets
    ## collecting deleted tweets ids and rate_limit_exceeded_tweets
    file.write('[\n')
    for tweet_id in dogs_df.tweet_id:
        try:
            
            dump_tweet_json(tweet_id, file)
            file.write(',\n')
        
        except tweepy.error.RateLimitError:
            rate_limit_exceeded_tweets.append(tweet_id)
        
        except Exception as e:
            deleted_tweets_ids.append(tweet_id)
            print(e,' ', tweet_id)
    
    ## Second Path of Querying tweets
    ## Making sure to collect rate_limit_exceeded_tweets if any 
    
    while rate_limit_exceeded_tweets:
        try:
            tweet_id = rate_limit_exceeded_tweets[0]
            dump_tweet_json(tweet_id, file)
            file.write(',\n')
            rate_limit_exceeded_tweets.pop(0)
            
        except tweepy.error.RateLimitError:
            time.sleep(60)
            
        except tweepy.error.TweepError as tr:
            if tr.message[0]['code'] == 144:
                rate_limit_exceeded_tweets.pop(0)
            else:
                print('Twitter Error: ', twitter_err)
        
        except Exception as e:
            print(e)
            
    file.seek(file.tell() - 3, os.SEEK_SET)
    file.write('\n]')

[{'code': 144, 'message': 'No status found with that ID.'}]   888202515573088257
[{'code': 144, 'message': 'No status found with that ID.'}]   873697596434513921
[{'code': 144, 'message': 'No status found with that ID.'}]   872668790621863937
[{'code': 144, 'message': 'No status found with that ID.'}]   872261713294495745
[{'code': 144, 'message': 'No status found with that ID.'}]   869988702071779329
[{'code': 144, 'message': 'No status found with that ID.'}]   866816280283807744
[{'code': 144, 'message': 'No status found with that ID.'}]   861769973181624320
[{'code': 144, 'message': 'No status found with that ID.'}]   856602993587888130
[{'code': 144, 'message': 'No status found with that ID.'}]   845459076796616705
[{'code': 144, 'message': 'No status found with that ID.'}]   844704788403113984
[{'code': 144, 'message': 'No status found with that ID.'}]   842892208864923648
[{'code': 144, 'message': 'No status found with that ID.'}]   837012587749474308
[{'code': 144, 'message': 'N

Rate limit reached. Sleeping for: 309


[{'code': 144, 'message': 'No status found with that ID.'}]   754011816964026368
[{'code': 144, 'message': 'No status found with that ID.'}]   680055455951884288


Rate limit reached. Sleeping for: 320


In [368]:
## Reading needed information from tweet_json to a dataframe
## needed info.: retweet_count, favorite_count
## in_reply_to_status_id, in_reply_to_user_id

df = pd.read_json('tweet_json.txt')
df.head(3)

contributors  coordinates          created_at display_text_range  \
0 NaN           NaN          2017-08-01 16:23:56  [0, 85]             
1 NaN           NaN          2017-08-01 00:17:27  [0, 138]            
2 NaN           NaN          2017-07-31 00:18:03  [0, 121]            

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  entities  \
0  {'hashtags': [], 'symbols': [], 'user_mentions': [], 'urls': [], 'media': [{'id': 892420639486877696, 'id_str': '892420639486877696', 'indices': [86, 109], 'media_url': 'http://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg', 'media_url_https': 'https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg', 'url': 'https://t.co/MgUWQ76dJU', 'display_url': 'pic.twitter.com/MgUWQ76dJU', 'expanded_url': 'https://twitter.com/dog_rates/status/892420643555336193/photo/1', 'type': 'photo', 'sizes': {'thumb': {'w': 150, 'h': 150, 'resize': 'crop'}, 'medium': {'w': 540, 'h': 528, 'resize': 'fit'}, 'small': {'w': 540, 'h': 528, 'resize': 'fit'}, 'large': {'w': 540, 'h': 528, 'resize': 'fit'}}}]}        
1  {'hashtags': [], 'symbols': [], 'user_mentions': [], 'urls': [], 'media': [{'id': 892177413194625024, 'id_str': '892177413194625024', 'indices': [139, 162], 'media_url': 'http://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg', 'media_url_https': 'https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg', 'url': 'https://t.co/0Xxu71qeIV', 'display_url': 'pic.twitter.com/0Xxu71qeIV', 'expanded_url': 'https://twitter.com/dog_rates/status/892177421306343426/photo/1', 'type': 'photo', 'sizes': {'thumb': {'w': 150, 'h': 150, 'resize': 'crop'}, 'medium': {'w': 1055, 'h': 1200, 'resize': 'fit'}, 'small': {'w': 598, 'h': 680, 'resize': 'fit'}, 'large': {'w': 1407, 'h': 1600, 'resize': 'fit'}}}]}   
2  {'hashtags': [], 'symbols': [], 'user_mentions': [], 'urls': [], 'media': [{'id': 891815175371796480, 'id_str': '891815175371796480', 'indices': [122, 145], 'media_url': 'http://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg', 'media_url_https': 'https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg', 'url': 'https://t.co/wUnZnhtVJB', 'display_url': 'pic.twitter.com/wUnZnhtVJB', 'expanded_url': 'https://twitter.com/dog_rates/status/891815181378084864/photo/1', 'type': 'photo', 'sizes': {'thumb': {'w': 150, 'h': 150, 'resize': 'crop'}, 'small': {'w': 510, 'h': 680, 'resize': 'fit'}, 'medium': {'w': 901, 'h': 1200, 'resize': 'fit'}, 'large': {'w': 1201, 'h': 1600, 'resize': 'fit'}}}]}    

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         extended_entities  \
0  {'media': [{'id': 892420639486877696, 'id_str': '892420639486877696', 'indices': [86, 109], 'media_url': 'http://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg', 'media_url_https': 'https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg', 'url': 'https://t.co/MgUWQ76dJU', 'display_url': 'pic.twitter.com/MgUWQ76dJU', 'expanded_url': 'https://twitter.com/dog_rat

In [369]:
df.columns

Index(['contributors', 'coordinates', 'created_at', 'display_text_range',
       'entities', 'extended_entities', 'favorite_count', 'favorited',
       'full_text', 'geo', 'id', 'id_str', 'in_reply_to_screen_name',
       'in_reply_to_status_id', 'in_reply_to_status_id_str',
       'in_reply_to_user_id', 'in_reply_to_user_id_str', 'is_quote_status',
       'lang', 'place', 'possibly_sensitive', 'possibly_sensitive_appealable',
       'quoted_status', 'quoted_status_id', 'quoted_status_id_str',
       'quoted_status_permalink', 'retweet_count', 'retweeted',
       'retweeted_status', 'source', 'truncated', 'user'],
      dtype='object')

In [370]:
## Checking completeness of wrangled parsed data and given data 
## After parsing the dataframe obs should be smaller than the given original data obs
## by just the deleted tweets

assert dogs_df.shape[0] == len(deleted_tweets_ids) + df.shape[0]

In [371]:
## Interested tweets info 

tweets_add_df = df[['retweet_count' , 'favorite_count', 'in_reply_to_status_id', 
                    'in_reply_to_user_id', 'is_quote_status', 'quoted_status_id', 'retweeted_status']]

tweets_add_df.head(3)

retweet_count  favorite_count  in_reply_to_status_id  in_reply_to_user_id  \
0  8193           37554          NaN                    NaN                    
1  6060           32278          NaN                    NaN                    
2  4008           24320          NaN                    NaN                    

   is_quote_status  quoted_status_id retweeted_status  
0  False           NaN                NaN              
1  False           NaN                NaN              
2  False           NaN                NaN

In [372]:
list(df['user'].apply(pd.Series))

['id',
 'id_str',
 'name',
 'screen_name',
 'location',
 'description',
 'url',
 'entities',
 'protected',
 'followers_count',
 'friends_count',
 'listed_count',
 'created_at',
 'favourites_count',
 'utc_offset',
 'time_zone',
 'geo_enabled',
 'verified',
 'statuses_count',
 'lang',
 'contributors_enabled',
 'is_translator',
 'is_translation_enabled',
 'profile_background_color',
 'profile_background_image_url',
 'profile_background_image_url_https',
 'profile_background_tile',
 'profile_image_url',
 'profile_image_url_https',
 'profile_banner_url',
 'profile_link_color',
 'profile_sidebar_border_color',
 'profile_sidebar_fill_color',
 'profile_text_color',
 'profile_use_background_image',
 'has_extended_profile',
 'default_profile',
 'default_profile_image',
 'following',
 'follow_request_sent',
 'notifications',
 'translator_type']

In [373]:
## Collecting info about users posting the tweets, most of them expected to be WeRateDogs page
## but if the tweet is a retweet, Different users shall be expected 

users_info = df['user'].apply(pd.Series)[['id', 'name', 'created_at', 'friends_count', 'followers_count', 'favourites_count', 'listed_count', 'statuses_count']]
users_info.head()

id         name                      created_at  friends_count  \
0  4196983835  WeRateDogs™  Sun Nov 15 21:41:29 +0000 2015  12              
1  4196983835  WeRateDogs™  Sun Nov 15 21:41:29 +0000 2015  12              
2  4196983835  WeRateDogs™  Sun Nov 15 21:41:29 +0000 2015  12              
3  4196983835  WeRateDogs™  Sun Nov 15 21:41:29 +0000 2015  12              
4  4196983835  WeRateDogs™  Sun Nov 15 21:41:29 +0000 2015  12              

   followers_count  favourites_count  listed_count  statuses_count  
0  8026597          141917            6187          10170           
1  8026597          141917            6187          10170           
2  8026597          141917            6187          10170           
3  8026597          141917            6187          10170           
4  8026597          141917            6187          10170

<hr>

## Assessing and Defining Issues in the Data

In [374]:
dogs_df.head(2)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193 NaN                    NaN                    
1  892177421306343426 NaN                    NaN                    

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   

                                                                               source  \
0  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                 text  \
0  This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU                                                        
1  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   

   retweeted_status_id  retweeted_status_user_id retweeted_status_timestamp  \
0 NaN                  NaN                        NaN                         
1 NaN                  NaN                        NaN                         

                                                     expanded_urls  \
0  https://twitter.com/dog_rates/status/892420643555336193/photo/1   
1  https://twitter.com/dog_rates/status/892177421306343426/photo/1   

   rating_numerator  rating_denominator     name doggo floofer pupper puppo  
0  13                10                  Phineas  None  None    None   None  
1  13                10                  Tilly    None  None    None   None

In [375]:
tweets_add_df.head()

retweet_count  favorite_count  in_reply_to_status_id  in_reply_to_user_id  \
0  8193           37554          NaN                    NaN                    
1  6060           32278          NaN                    NaN                    
2  4008           24320          NaN                    NaN                    
3  8347           40905          NaN                    NaN                    
4  9040           39097          NaN                    NaN                    

   is_quote_status  quoted_status_id retweeted_status  
0  False           NaN                NaN              
1  False           NaN                NaN              
2  False           NaN                NaN              
3  False           NaN                NaN              
4  False           NaN                NaN

In [376]:
predict_df.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0  1        Welsh_springer_spaniel  0.465074  True    collie               
1  1        redbone                 0.506826  True    miniature_pinscher   
2  1        German_shepherd         0.596461  True    malinois             
3  1        Rhodesian_ridgeback     0.408143  True    redbone              
4  1        miniature_pinscher      0.560311  True    Rottweiler           

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665  True    Shetland_sheepdog    0.061428  True    
1  0.074192  True    Rhodesian_ridgeback  0.072010  True    
2  0.138584  True    bloodhound           0.116197  True    
3  0.360687  True    miniature_pinscher   0.222752  True    
4  0.243682  True    Doberman             0.154629  True

### Assessing dogs_df

> In this section, I will assess the data. In every step, I may find an issue and will document it in markdown cells, 
at the end of the section, all issues will be summarized and classified.

In [377]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_colwidth', -1)

In [378]:
dogs_df.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193 NaN                    NaN                    
1  892177421306343426 NaN                    NaN                    
2  891815181378084864 NaN                    NaN                    
3  891689557279858688 NaN                    NaN                    
4  891327558926688256 NaN                    NaN                    

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                                                               source  \
0  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
3  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
4  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                 text  \
0  This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU                                                        
1  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
2  This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB                    
3  This is Darla. She commenced a snooze mid meal. 13/10 happens to the best of us https://t.co/tD36da7qLQ                                                              
4  This is Franklin. He would like you to stop calling him "cute." He is a very fierce shark and should be respected as such. 12/10 #BarkWeek https://t.co/AtUZn91f7f   

   retweeted_status_id  retweeted_status_user_id retweeted_status_timestamp  \
0 NaN                  NaN                        NaN                         
1 NaN                  NaN                        NaN                         
2 NaN                  NaN                        NaN                         
3 NaN                  NaN                        NaN                         
4 NaN                  NaN                        NaN                         

                                                                                                                     expanded_urls  \
0  https://twitter.com/dog_rates/status/892420643555336193/photo/1                                                                   
1  https://twitter.com/dog_rates/status/892177421306343426/photo/1                                                                   
2  https://twitter.com/dog_rates/status/891815181378084864/photo/1                                                                   
3  https://twitter.com/dog_rates/status/891689557279858688/photo/1                                                                   
4  https://twitter.com/dog_rates/status/891327558926688256/photo/1,https://twitter.com/dog_rates/status/891327558926688256/photo/1   

   rating_numerator  rating_denominator      name doggo floofer pupper puppo  
0  13                10                  Phineas   None  None    None   None  
1  13                10                  Tilly     None  None    None   None  
2  12                10                  Archie    None  None    None   None  
3  13                10                  Darla     None  None    None   None  
4  12                10                  Franklin  None  None    None   None

In [379]:
dogs_df.tail()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
2351  666049248165822465 NaN                    NaN                    
2352  666044226329800704 NaN                    NaN                    
2353  666033412701032449 NaN                    NaN                    
2354  666029285002620928 NaN                    NaN                    
2355  666020888022790149 NaN                    NaN                    

                      timestamp  \
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                                                  source  \
2351  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2352  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2353  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2354  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2355  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                             text  \
2351  Here we have a 1949 1st generation vulpix. Enjoys sweat tea and Fox News. Cannot be phased. 5/10 https://t.co/4B7cOc1EDq                      
2352  This is a purebred Piers Morgan. Loves to Netflix and chill. Always looks like he forgot to unplug the iron. 6/10 https://t.co/DWnyCjf2mx     
2353  Here is a very happy pup. Big fan of well-maintained decks. Just look at that tongue. 9/10 would cuddle af https://t.co/y671yMhoiR            
2354  This is a western brown Mitsubishi terrier. Upset about leaf. Actually 2 dogs here. 7/10 would walk the shit out of https://t.co/r7mOb2m0UI   
2355  Here we have a Japanese Irish Setter. Lost eye in Vietnam (?). Big fan of relaxing on stair. 8/10 would pet https://t.co/BLDqew2Ijj           

      retweeted_status_id  retweeted_status_user_id  \
2351 NaN                  NaN                         
2352 NaN                  NaN                         
2353 NaN                  NaN                         
2354 NaN                  NaN                         
2355 NaN                  NaN                         

     retweeted_status_timestamp  \
2351  NaN                         
2352  NaN                         
2353  NaN                         
2354  NaN                         
2355  NaN                         

                                                        expanded_urls  \
2351  https://twitter.com/dog_rates/status/666049248165822465/photo/1   
2352  https://twitter.com/dog_rates/status/666044226329800704/photo/1   
2353  https://twitter.com/dog_rates/status/666033412701032449/photo/1   
2354  https://twitter.com/dog_rates/status/666029285002620928/photo/1   
2355  https://twitter.com/dog_rates/status/666020888022790149/photo/1   

      rating_numerator  rating_denominator  name doggo floofer pupper puppo  
2351  5                 10                  None  None  None    None   None  
2352  6                 10                  a     None  None    None   None  
2353  9                 10                  a     None  None    None   None  
2354  7                 10                  a     None  None    None   None  
2355  8                 10                  None  None  None    None   None

In [380]:
dogs_df.sample(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
467   817120970343411712 NaN                    NaN                    
1793  677331501395156992 NaN                    NaN                    
1776  677961670166224897 NaN                    NaN                    
2020  672095186491711488 NaN                    NaN                    
927   754856583969079297 NaN                    NaN                    

                      timestamp  \
467   2017-01-05 21:29:55 +0000   
1793  2015-12-17 03:36:28 +0000   
1776  2015-12-18 21:20:32 +0000   
2020  2015-12-02 16:49:14 +0000   
927   2016-07-18 01:53:28 +0000   

                                                                                  source  \
467   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1793  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1776  <a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>                      
2020  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
927   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                             text  \
467   Name a more iconic quartet... I'll wait. 13/10 for all https://t.co/kCLgD8687T                                                                
1793  Meet Humphrey. He's a Northern Polyp Viagra. One ear works. Face stuck like that. Always surprised. 9/10 petable af https://t.co/FS7eJQM2F4   
1776  This is Izzy. She's showing off the dance moves she's been working on. 11/10 I guess hard work pays off https://t.co/4JS92YAxTi               
2020  This is Tuco. That's the toast that killed his father. 9/10 https://t.co/ujnWy26RMe                                                           
927   This is Winnie. She's not a fan of the fast moving air. 11/10 objects in mirror may be more fluffy than they appear https://t.co/FyHrk20gUR   

      retweeted_status_id  retweeted_status_user_id  \
467  NaN                  NaN                         
1793 NaN                  NaN                         
1776 NaN                  NaN                         
2020 NaN                  NaN                         
927  NaN                  NaN                         

     retweeted_status_timestamp  \
467   NaN                         
1793  NaN                         
1776  NaN                         
2020  NaN                         
927   NaN                         

                                                                                                                                                                                        expanded_urls  \
467   https://twitter.com/dog_rates/status/817120970343411712/photo/1                                                                                                                                   
1793  https://twitter.com/dog_rates/status/677331501395156992/photo/1                                                                                                                                   
1776  https://vine.co/v/iKuMDuYV0aZ                                                                                                                                                                     
2020  https://twitter.com/dog_rates/status/672095186491711488/photo/1                                                                                                                                   
927   https://twitter.com/dog_rates/status/754856583969079297/photo/1,https://twitter.com/dog_rates/status/754856583969079297/photo/1,https://twitter.com/dog_rates/status/754856583969079297/photo/1   

      rating_numerator  rating_denominator      name doggo floofer pupper  \
467   13                10                  None      None  None    None    
1793  9              

In [381]:
dogs_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

From visual assessment and info() function:
* the source column contains an html link tag that contians three types of information in one column, the source of the tweet, link of the source, 'rel' attribute.
* The doggo, pupper and puppo are all possible dog stage, so it's better  of to represent them by one column e.g dog_stage column and the stage_column should be categorized.
* Column doggo, pupper, and puppo sometimes are all Nones while a dog should be really in one of the three stages or sometimes big doggo is considered woofer, but here in the system, it seems dogs are categorized between only pupper, doggo and puppo and from further investigation visual assessment, it seems that the category of the dog is extracted from the text, so when the tweet text doesn't contain any of these words or use different abbreviations like 'pup', the values are none.
* Not all ratings numerators are higher than 10, some ratings actually below 10, from visual assessment, and seeing the dates, it seems all in this small sample posted in 2015, more programmatic assessment needed here.
* The (in_reply_to_status_id, in_reply_to_user_id), (retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp)	seems to contain many Nones, as we see from the info function, just 78 non-nulls for in_reply and 181 for retweets and that might not come as a surprise since most tweets are originally posted by the page, and such official rating pages may not retweet often, rather a personal twitter may contain many retweets.
* The doggo, floofer, pupper, puppo doesn't contain any null entries from the info() function although we seem many Nones from the visual assessment and could be that string 'None' is inserted in them not object None.
* expanded_urls contains 59 null objects that means there are 59 tweets with no url, this data could be gathered from the json file, if exists.
* There is a 'None' names in the tail and 'a' names, too. Checking the text, it seems names are extracted from statements in the form 'This is name', 'Here is name', if anything came in place of name, it's would be the name, and if the pattern doesn' exist, it will be None. A better text pattern may be used, but from this sample head, tail, sample, Sometimes names aren't really stated, somtimes breed, type or origin is stated, so we should furhter investigate name column.
* 'tweet_id' is integer, 'in_reply_to_status_id', 'retweeted_status_user_id', 'retweeted_status_id', 'retweeted_status_user_id' are floats. Finally, timestamp columns are string objects not datetime.
* Finally, the 'text' columns contains two pieces of information both the text and the t.co url link wrappers. So this is a tidyness issue. Moreover, from visual assessment and checking url, t.co links usually refer to the same page as the expaned_url, but this I will check later

Next I will check for values of every interesting column to make sure case of 'None' strings doesn't happen to other columns

In [382]:
def value_counts_per_column(df, columns):
    """
    Takes a dataframe and a set of columns in the dataframe 
    and prints out the value_counts of each column
    """
    for column in columns:
        print(df[column].value_counts())
        print('\n\n', '==' * 20, '\n\n')
        

In [383]:
list(dogs_df)

['tweet_id',
 'in_reply_to_status_id',
 'in_reply_to_user_id',
 'timestamp',
 'source',
 'text',
 'retweeted_status_id',
 'retweeted_status_user_id',
 'retweeted_status_timestamp',
 'expanded_urls',
 'rating_numerator',
 'rating_denominator',
 'name',
 'doggo',
 'floofer',
 'pupper',
 'puppo']

In [384]:
value_counts_per_column(dogs_df, ['name', 'doggo', 'floofer', 'pupper', 'puppo', 'source'])

None            745
a               55 
Charlie         12 
Lucy            11 
Oliver          11 
Cooper          11 
Tucker          10 
Lola            10 
Penny           10 
Winston         9  
Bo              9  
the             8  
Sadie           8  
Bailey          7  
Daisy           7  
Buddy           7  
an              7  
Toby            7  
Koda            6  
Rusty           6  
Oscar           6  
Scout           6  
Stanley         6  
Jack            6  
Jax             6  
Dave            6  
Milo            6  
Bella           6  
Leo             6  
Larry           5  
Gus             5  
very            5  
Alfie           5  
Louis           5  
Sammy           5  
George          5  
Phil            5  
Finn            5  
Bentley         5  
Sunny           5  
Chester         5  
Oakley          5  
Boomer          4  
Reggie          4  
Clarence        4  
Archie          4  
Clark           4  
Loki            4  
Luna            4  
Hank            4  


**Observations per column(s):**
* 'name': There are many Nones and other unreasonable names like 'a', 'an', 'old', 'the'
* 'doggo', 'floofer', 'pupper', 'puppo': Values are either None or their column name, 'None' are dominant in those columns. 
* 'source': is either from Twitter for iPhone app, Vine, the Web or TweetDeck, we can see that the rel attribute takes always 'nofollow' value, so actually its existence when tidying up the column may be irrelevant, more space for no benefit, we make make a note that all links are 'nofollow' for browsers. Hence, I may argue that the html attr 'rel' is not important in our investigation or representation of the data-set.

In [385]:
## Further investigating name column

pd.set_option('display.max_rows', 500)

dogs_df['name'].value_counts()[dogs_df['name'].value_counts() > 3]

None        745
a           55 
Charlie     12 
Lucy        11 
Oliver      11 
Cooper      11 
Tucker      10 
Lola        10 
Penny       10 
Winston     9  
Bo          9  
the         8  
Sadie       8  
Bailey      7  
Daisy       7  
Buddy       7  
an          7  
Toby        7  
Koda        6  
Rusty       6  
Oscar       6  
Scout       6  
Stanley     6  
Jack        6  
Jax         6  
Dave        6  
Milo        6  
Bella       6  
Leo         6  
Larry       5  
Gus         5  
very        5  
Alfie       5  
Louis       5  
Sammy       5  
George      5  
Phil        5  
Finn        5  
Bentley     5  
Sunny       5  
Chester     5  
Oakley      5  
Boomer      4  
Reggie      4  
Clarence    4  
Archie      4  
Clark       4  
Loki        4  
Luna        4  
Hank        4  
Winnie      4  
Maddie      4  
one         4  
Shadow      4  
Walter      4  
Maggie      4  
just        4  
Ruby        4  
quite       4  
Sampson     4  
Bruce       4  
Dexter      4  
Gerald  

<hr>

**Investigating name column furhter we notice those issues:**
1. very high count of None (745)
2. Articles like 'a', 'an', 'the' adverbs like 'just', 'very', 'quite', adj like 'old' and others like  'one'.
3. Some Strange names like Chip Scooter Moose Bentley Shadow, but names by nature aren't justified, so I will dig further to those specific names, but most likely they won't considered quality issues.

The articles, adverbs and others like 'one' are all begin with lower case letters, that might help cleaning those data instead of getting list of all possible articles, adverbs and pronouns to check their existence in the names column

<hr>

In [386]:
def any_None_empty_object(df):
    """
    checks there is any 'None' or empty '' string objects values in string 
    columns and prints out the column name if any
    """
    for column in list(df):
        if df[column].dtype == np.object:
            if any(df[column].str.lower() == 'none'):
                print(column, "'none' strings exists")
            elif any(df[column].str.strip() == ''):
                print(column, 'Empty strings exist')

In [387]:
any_None_empty_object(dogs_df)

name 'none' strings exists
doggo 'none' strings exists
floofer 'none' strings exists
pupper 'none' strings exists
puppo 'none' strings exists


<hr>

It seems there is no empty string objects for any string column and srings containing 'none' strings are name, doggo, pupper, puppo which we have assessed before.

<hr>

In [388]:
pd.set_option('display.max_rows', 10)

In [389]:
## Additional assessment for remaining text columns

value_counts_per_column(dogs_df, ['timestamp', 'text', 'expanded_urls', 'retweeted_status_timestamp'])

2016-03-05 04:17:02 +0000    1
2015-12-07 02:21:29 +0000    1
2015-11-19 03:52:34 +0000    1
2015-12-03 02:21:48 +0000    1
2017-01-22 01:12:59 +0000    1
                            ..
2016-02-04 02:00:27 +0000    1
2016-03-10 01:35:01 +0000    1
2016-02-06 03:50:33 +0000    1
2016-02-25 16:53:11 +0000    1
2015-11-28 19:04:19 +0000    1
Name: timestamp, Length: 2356, dtype: int64




This is Orion. He just got back from the dentist. Cavity free af. 12/10 would give extra pats https://t.co/Y4DZx2UWsr                           1
This is Ike. He's demonstrating the pupmost restraint. 13/10 super good boy https://t.co/6gHoGah9nm                                             1
This is Daisy. She's puppears to be rare as all h*ck. Only seven like her currently domesticated. 13/10 pettable af https://t.co/meUc8jufAO     1
"Don't ever talk to me or my son again." ...both 10/10 https://t.co/b8ncwl6TlE                                                                  1
Meet Rilo. He's a Northern 

**Observations:**
* Timestamp objects seems to be unique and all follow same format.
* Texts also tend to be unique as there is no text with value_count more than 1
* Expaned urls have some urls with value count more than one while we know every tweet has unique tweet_id. 
<br>In addition, For some entries, expanded urls have the same url repeated e.g. "https://twitter.com/dog_rates/status/782305867769217024/photo/1,
https://twitter.com/dog_rates/status/782305867769217024/photo/1,
https://twitter.com/dog_rates/status/782305867769217024/photo/1" In addition, the urls seems to be the same, but this needs to be checked further to figure out why there are multiple links, and why the links are the same and whether the links are the same for all expaneded_urls containing more than one url. Based on those questoins, this be a quality issue or a tidiness issue or both to be considered in the cleaning.

<hr>

In [390]:
## Checking uniqueness of tweet_id and that there is no None values
assert dogs_df['tweet_id'].nunique() == dogs_df.shape[0]

In [391]:
dogs_df.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03  7.800000e+01           7.800000e+01          
mean   7.427716e+17  7.455079e+17           2.014171e+16          
std    6.856705e+16  7.582492e+16           1.252797e+17          
min    6.660209e+17  6.658147e+17           1.185634e+07          
25%    6.783989e+17  6.757419e+17           3.086374e+08          
50%    7.196279e+17  7.038708e+17           4.196984e+09          
75%    7.993373e+17  8.257804e+17           4.196984e+09          
max    8.924206e+17  8.862664e+17           8.405479e+17          

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count  1.810000e+02         1.810000e+02              2356.000000        
mean   7.720400e+17         1.241698e+16              13.126486          
std    6.236928e+16         9.599254e+16              45.876648          
min    6.661041e+17         7.832140e+05              0.000000           
25%    7.186315e+17         4.196984e+09              10.000000          
50%    7.804657e+17         4.196984e+09              11.000000          
75%    8.203146e+17         4.196984e+09              12.000000          
max    8.874740e+17         7.874618e+17              1776.000000        

       rating_denominator  
count  2356.000000         
mean   10.455433           
std    6.745237            
min    0.000000            
25%    10.000000           
50%    10.000000           
75%    10.000000           
max    170.000000

**Observations:**
* Checking (min, 25%, 50%, 75%, max), every tweet id (tweet_id, in_reply_to_status_id, retweeted_status_id) tend to be 18 digits, hence a string representation here is meaningful as mean, std doens't really mean anything here.
* On the other hand, user ids (in_reply_to_user_id, retweeted_status_user_id) doesn't all have 18 digits, some have 8 or 6 digits, up to 18 digits, according to the article [Interpreting Twitter id](https://www.adweek.com/digital/twitter-user-id/) twitter user ids gave an ordering of when a user account was created, but thinking of the 18 digit user ids, it seems the incrementing or incrementing fashion like user id system has been replaced by 18 digits, that's just a guess, to make sure we should investigate user ids and their account launch dates. In addition, this guess is based on the fact that there is supposedly about 330 million twitter user today and there is about 7.5 billion people in the world as whole, 18 digits is much much higher number, meaning even if every person on Earth is registered on twitter, every one shall have about a billion twitter accounts which is untrue.
* Rating numerators have untrue or outlier values like 1776, 0 biasing the mean and std. (quality issue)
* Rating denomenators have untrue or outlier values like 0, 170 biasing while all denomenators should be supposedly 10. (quality issue)

> **Summarizing Data issues in dogs_df:**
* **Tidyness Issues: **
       1. Converting source column HTML to its constituent parts url, and app source, 'rel' will be discarded since it's alwasy 'no-follow'.
       2. Converting pupper, puppo, doggo value columns to be values in one column stage_column.
* **Quality Issues: **
       1. 'name' column: Clearing 'None' names and inappropriate false names like 'a', 'an' and 'the'.
       2. Changing 'floofer' column to boolean.

### Cleaning dogs_df

In [401]:
dogs_clean = dogs_df.copy()

First, I will convert tweeets ids to string for easier visual assessment and checking for ids and because int and float ids values doesn't convey any numerical ordering for ids.

In [402]:
## Converting tweet_id to string
dogs_clean.tweet_id = dogs_clean.tweet_id.astype(str)

When converting float ids to string, I will first need to convert them to integers, then to str, the problem is with Nan when converting from float to int, so first I will first nans with -1, then after conversion they will be returend to Nan again.

In [403]:
# Converting float ids to string
for column in list(dogs_clean):
    if column[-3:] == '_id' and any(pd.isnull(dogs_clean[column])):
        dogs_clean[column].fillna(-1, inplace = True)
        dogs_clean[column]  = dogs_clean[column].astype(np.int64).astype(str)
        dogs_clean.loc[dogs_clean[column] == '-1', column] = None
    
dogs_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null object
in_reply_to_status_id         78 non-null object
in_reply_to_user_id           78 non-null object
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null object
retweeted_status_user_id      181 non-null object
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: int64(2), object(15)
memory

Cleaning Tidiness Issues:
1. Text, url 
2. source

### Source column

In [395]:
dogs_clean.source.head()

0    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>
1    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>
2    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>
3    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>
4    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>
Name: source, dtype: object

In [396]:
dogs_clean[['source_app', 'source_url']] = dogs_clean.source.str.extract('<a href="(.*)" rel="nofollow">(.*)</a>')[[1,0]]

In [397]:
dogs_clean[['source', 'source_app', 'source_url']].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 3 columns):
source        2356 non-null object
source_app    2356 non-null object
source_url    2356 non-null object
dtypes: object(3)
memory usage: 55.3+ KB


In [398]:
dogs_clean[['source', 'source_app', 'source_url']].sample(5)

source  \
2034  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1905  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1007  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1653  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
801   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

              source_app                          source_url  
2034  Twitter for iPhone  http://twitter.com/download/iphone  
1905  Twitter for iPhone  http://twitter.com/download/iphone  
1007  Twitter for iPhone  http://twitter.com/download/iphone  
1653  Twitter for iPhone  http://twitter.com/download/iphone  
801   Twitter for iPhone  http://twitter.com/download/iphone

In [399]:
dogs_clean.drop(columns = ['source'], inplace=True)

In [400]:
list(dogs_clean)

['tweet_id',
 'in_reply_to_status_id',
 'in_reply_to_user_id',
 'timestamp',
 'text',
 'retweeted_status_id',
 'retweeted_status_user_id',
 'retweeted_status_timestamp',
 'expanded_urls',
 'rating_numerator',
 'rating_denominator',
 'name',
 'doggo',
 'floofer',
 'pupper',
 'puppo',
 'source_app',
 'source_url']

### text, url

Since the urls are [t.co](https://help.twitter.com/en/using-twitter/how-to-tweet-a-link) shortened twitter url, the regex here is specific to capture those specific urls

In [309]:
# Check this if there is time for Nones
# dogs_clean.iloc[2298]

In [310]:
dogs_clean['shortened_urls'] = dogs_clean.text.str.extract('(https?://t.co/[a-zA-Z0-9]{10})')[0]

In [311]:
dogs_clean.text.head()

0    This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU                                                     
1    This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV
2    This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB                 
3    This is Darla. She commenced a snooze mid meal. 13/10 happens to the best of us https://t.co/tD36da7qLQ                                                           
4    This is Franklin. He would like you to stop calling him "cute." He is a very fierce shark and should be respected as such. 12/10 #BarkWeek https://t.co/AtUZn91f7f
Name: text, dtype: object

In [312]:
## Removing urls from text 
dogs_clean.text = dogs_clean.text.str.replace('(https?://t.co/[a-zA-Z0-9]{10})', '')
dogs_clean.text.head()

0    This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10                                                      
1    This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 
2    This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10                  
3    This is Darla. She commenced a snooze mid meal. 13/10 happens to the best of us                                                            
4    This is Franklin. He would like you to stop calling him "cute." He is a very fierce shark and should be respected as such. 12/10 #BarkWeek 
Name: text, dtype: object

In [313]:
dogs_clean[['shortened_urls', 'expanded_urls', 'text']].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 3 columns):
shortened_urls    2213 non-null object
expanded_urls     2297 non-null object
text              2356 non-null object
dtypes: object(3)
memory usage: 55.3+ KB


Now we have separated text and urls, it seems some text doesn't contain shortenered_urls, hence I will iterate over those

In [314]:
##  Removing the urls from text
none_shortened_urls = dogs_clean[pd.isnull(dogs_clean.shortened_urls)]
none_shortened_urls.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 143 entries, 30 to 2298
Data columns (total 18 columns):
tweet_id                      143 non-null object
in_reply_to_status_id         55 non-null object
in_reply_to_user_id           55 non-null object
timestamp                     143 non-null object
source                        143 non-null object
text                          143 non-null object
retweeted_status_id           85 non-null object
retweeted_status_user_id      85 non-null object
retweeted_status_timestamp    85 non-null object
expanded_urls                 84 non-null object
rating_numerator              143 non-null int64
rating_denominator            143 non-null int64
name                          143 non-null object
doggo                         143 non-null object
floofer                       143 non-null object
pupper                        143 non-null object
puppo                         143 non-null object
shortened_urls                0 non-null object
dty

In [316]:
none_shortened_urls.head()

tweet_id in_reply_to_status_id in_reply_to_user_id  \
30  886267009285017600  886266357075128320    2281181600           
55  881633300179243008  881607037314052096    47384430             
64  879674319642796034  879553827334172672    3105440746           
68  879130579576475649  None                  None                 
73  878404777348136964  None                  None                 

                    timestamp  \
30  2017-07-15 16:51:35 +0000   
55  2017-07-02 21:58:53 +0000   
64  2017-06-27 12:14:36 +0000   
68  2017-06-26 00:13:58 +0000   
73  2017-06-24 00:09:53 +0000   

                                                                                source  \
30  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
55  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
64  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
68  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
73  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                              text  \
30  @NonWhiteHat @MayhewMayhem omg hello tanner you are a scary good boy 12/10 would pet with extreme caution                                        
55  @roushfenway These are good dogs but 17/10 is an emotional impulse rating. More like 13/10s                                                      
64  @RealKentMurphy 14/10 confirmed                                                                                                                  
68  RT @dog_rates: This is Emmy. She was adopted today. Massive round of pupplause for Emmy and her new family. 14/10 for all involved https://…     
73  RT @dog_rates: Meet Shadow. In an attempt to reach maximum zooming borkdrive, he tore his ACL. Still 13/10 tho. Help him out below\n\nhttps:/…   

   retweeted_status_id retweeted_status_user_id retweeted_status_timestamp  \
30  None                None                     NaN                         
55  None                None                     NaN                         
64  None                None                     NaN                         
68  878057613040115712  4196983835               2017-06-23 01:10:23 +0000   
73  878281511006478336  4196983835               2017-06-23 16:00:04 +0000   

                                                                                                                      expanded_urls  \
30  NaN                                                                                                                               
55  NaN                                                                                                                               
64  NaN                                                                                                                               
68  https://twitter.com/dog_rates/status/878057613040115712/photo/1,https://twitter.com/dog_rates/status/878057613040115712/photo/1   
73  https://www.gofundme.com/3yd6y1c,https://twitter.com/dog_rates/status/878281511006478336/photo/1                                  

    rating_numerator  rating_denominator    name doggo floofer pupper puppo  \
30  12                10                  None    None  None    None   None   
55  17                10                  None    None  None    None   None   
64  14                10                  None    None  None    None   None   
68  14                10                  Emmy    None  None    None   None   
73  13                10                  Shadow  None  None    None   None   

   shortened_urls  
30  NaN            
55  NaN            
64  NaN            
68  NaN            
73  NaN

In [317]:
none_shortened_urls.tail()

tweet_id in_reply_to_status_id in_reply_to_user_id  \
1940  673716320723169284  673715861853720576    4196983835           
2038  671550332464455680  671544874165002240    4196983835           
2149  669684865554620416  669354382627049472    4196983835           
2189  668967877119254528  668920717132582912    21435658             
2298  667070482143944705  667065535570550784    4196983835           

                      timestamp  \
1940  2015-12-07 04:11:02 +0000   
2038  2015-12-01 04:44:10 +0000   
2149  2015-11-26 01:11:28 +0000   
2189  2015-11-24 01:42:25 +0000   
2298  2015-11-18 20:02:51 +0000   

                                                                                  source  \
1940  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2038  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2149  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2189  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2298  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                         text  \
1940  The millennials have spoken and we've decided to immediately demote to a 1/10. Thank you                                                  
2038  After 22 minutes of careful deliberation this dog is being demoted to a 1/10. The longer you look at him the more terrifying he becomes   
2149  After countless hours of research and hundreds of formula alterations we have concluded that Dug should be bumped to an 11/10             
2189  12/10 good shit Bubka\n@wane15                                                                                                            
2298  After much debate this dog is being upgraded to 10/10. I repeat 10/10                                                                     

     retweeted_status_id retweeted_status_user_id retweeted_status_timestamp  \
1940  None                None                     NaN                         
2038  None                None                     NaN                         
2149  None                None                     NaN                         
2189  None                None                     NaN                         
2298  None                None                     NaN                         

     expanded_urls  rating_numerator  rating_denominator  name doggo floofer  \
1940  NaN           1                 10                  None  None  None     
2038  NaN           1                 10                  None  None  None     
2149  NaN           11                10                  None  None  None     
2189  NaN           12                10                  None  None  None     
2298  NaN           10                10                  None  None  None     

     pupper puppo shortened_urls  
1940  None   None  NaN            
2038  None   None  NaN            
2149  None   None  NaN            
2189  None   None  NaN            
2298  None   None  NaN

In [318]:
none_shortened_urls.sample()

tweet_id in_reply_to_status_id in_reply_to_user_id  \
1895  674742531037511680  674739953134403584    4196983835           

                      timestamp  \
1895  2015-12-10 00:08:50 +0000   

                                                                                  source  \
1895  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                             text  \
1895  Some clarification is required. The dog is singing Cher and that is more than worthy of an 11/10. Thank you   

     retweeted_status_id retweeted_status_user_id retweeted_status_timestamp  \
1895  None                None                     NaN                         

     expanded_urls  rating_numerator  rating_denominator  name doggo floofer  \
1895  NaN           11                10                  None  None  None     

     pupper puppo shortened_urls  
1895  None   None  NaN

We can see that some of those null shortened_urls tweets have null expanded_urls, too. In this case, all of the processed tweets through head, tail and sample tend to be a reply tweet as the in_reply_to_status_id, in_reply_to_user_id columns aren't empty.<br>
On the other hand, some of those null shortened_urls tweets do have expaned_urls values and checking through head, tail, and sample, those tweets tend to be retweets.<br>
**That said, some data issues has been found here:**
* @user appended at the begining of those in reply texts .e.g @NonWhiteHat @MayhewMayhem 
* RT @user: at the beginning of each retweet text appened to retweet texts .e.g RT @dog_rates:, checking the twitter [retweet doc intro](https://developer.twitter.com/en/docs/tweets/data-dictionary/overview/intro-to-tweet-json.html), it seems this appending actoin is deliberately done by the twitter api itself. 
* Truncated urls at the end in retweet texts (.e.g  'https://…', 'https:/…', 'https://t.co/…')
* Expanded urls do have different urls sometimes, for example https://www.gofundme.com/3yd6y1c, https://twitter.com/dog_rates/status/878281511006478336/photo/1 at the end of the head and the first link is a link included by the WeRateDogs tweet for funding, so we may expect that expanded tweets include also links included in the tweet
* Finally, we may retain in replies url by the following link format http://twitter.com/{twitter-user-id}/status/{tweet-status-id}, for example https://twitter.com/dog_rates/status/886267009285017600.

<hr>

fixing puppo, pupper, doggo columns

Before converting puppo, pupper, doggo to a stage column, we should clean the 'None' string objects

In [145]:
def cleaning_None_objects(df, columns):
    """
    Converting string 'None' object to None
    """
    for column in columns:
        df.loc[df[column] == 'None', column] = None
        

In [148]:
cleaning_None_objects(dogs_clean, ['doggo', 'pupper', 'puppo', 'floofer'])

In [152]:
dogs_clean[['doggo', 'puppo', 'pupper']].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 3 columns):
doggo     97 non-null object
puppo     30 non-null object
pupper    257 non-null object
dtypes: object(3)
memory usage: 55.3+ KB


Also, it's logical that a dog cannot be both pupper and doggo for example because they're different ages, but nothing is guaranteed from messy data, so I will check for this

In [153]:
## Checking for crossed values before melting

# this function checks if there dogs classified as either (doggo, pupper), 
# (doggo, puppo), (pupper, puppo) or (doggo, pupper, puppo)

dogs_stages = ['doggo', 'puppo', 'pupper']

for i in range(len(dogs_stages)):
    j = i + 1
    for j in range(i + 1, len(dogs_stages)):
        print('(%s, %s)' %(dogs_stages[i], dogs_stages[j]),
             dogs_clean.query('%s == "%s" and %s == "%s"'%(dogs_stages[i], dogs_stages[i], dogs_stages[j], dogs_stages[j])).shape[0]
             )

## Checking for three columns
print('(%s, %s, %s)' %('doggo', 'puppo', 'pupper'),
             dogs_clean.query('doggo == "doggo" and puppo == "puppo" and pupper == "pupper"').shape[0]
     )

(doggo, puppo) 1
(doggo, pupper) 12
(puppo, pupper) 0
(doggo, puppo, pupper) 0


It seems that there are 1 dog classifed as both doggo and puppo and 12 classifed as both doggo and pupper, so I will check the urls of those dogs

In [154]:
dogs_clean.query('doggo == "doggo" and puppo == "puppo"')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
191  855851453814013952 NaN                    NaN                    

                     timestamp  \
191  2017-04-22 18:31:02 +0000   

                                                                                 source  \
191  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                               text  \
191  Here's a puppo participating in the #ScienceMarch. Cleverly disguising her own doggo agenda. 13/10 would keep the planet habitable for https://t.co/cMhq16isel   

     retweeted_status_id  retweeted_status_user_id retweeted_status_timestamp  \
191 NaN                  NaN                        NaN                         

                                                       expanded_urls  \
191  https://twitter.com/dog_rates/status/855851453814013952/photo/1   

     rating_numerator  rating_denominator  name  doggo floofer pupper  puppo  
191  13                10                  None  doggo  None    None   puppo

It seems that this dog is actually puppo, so I will change this doggo entry to None

In [155]:
dogs_clean.loc[191, 'doggo'] = None

In [157]:
## Testing
assert dogs_clean.query('doggo == "doggo" and puppo == "puppo"').shape[0] == 0

In [160]:
pd.set_option('display.max_rows', 30)
dogs_clean.query('doggo == "doggo" and pupper == "pupper"')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
460   817777686764523521 NaN                    NaN                    
531   808106460588765185 NaN                    NaN                    
565   802265048156610565  7.331095e+17           4.196984e+09          
575   801115127852503040 NaN                    NaN                    
705   785639753186217984 NaN                    NaN                    
733   781308096455073793 NaN                    NaN                    
778   775898661951791106 NaN                    NaN                    
822   770093767776997377 NaN                    NaN                    
889   759793422261743616 NaN                    NaN                    
956   751583847268179968 NaN                    NaN                    
1063  741067306818797568 NaN                    NaN                    
1113  733109485275860992 NaN                    NaN                    

                      timestamp  \
460   2017-01-07 16:59:28 +0000   
531   2016-12-12 00:29:28 +0000   
565   2016-11-25 21:37:47 +0000   
575   2016-11-22 17:28:25 +0000   
705   2016-10-11 00:34:48 +0000   
733   2016-09-29 01:42:20 +0000   
778   2016-09-14 03:27:11 +0000   
822   2016-08-29 03:00:36 +0000   
889   2016-07-31 16:50:42 +0000   
956   2016-07-09 01:08:47 +0000   
1063  2016-06-10 00:39:48 +0000   
1113  2016-05-19 01:38:16 +0000   

                                                                                  source  \
460   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
531   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
565   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
575   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
705   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
733   <a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>                      
778   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
822   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
889   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
956   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1063  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1113  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                      text  \
460   This is Dido. She's playing the lead role in "Pupper Stops to Catch Snow Before Resuming Shadow Box with Dried Apple." 13/10 (IG: didodoggo) https://t.co/m7isZrOBX7   
531   Here we have Burke (pupper) and Dexter (doggo). Pupper wants to be exactly like doggo. Both 12/10 would pet at same time https://t.co/ANBpEYHaho                       
565   Like doggo, like pupper version 2. Both 11/10 https://t.co/9IxWAXFqze                                                                                                  
575   This is Bones. He's being haunted by another doggo of roughly the same size. 12/10 deep breaths pupper everything's fine https://t.co/55Dqe0SJNj                       
705   This is Pinot. He's a sophisticated doggo. You can tell by the hat. Also pointier than your average pupper. Still 10/10 would pet cautiously https://t.co/f2wmLZTPHd   
733   Pupper butt 1, Doggo 0. Both 12/10 https://t.co/WQvcPEpH2u                                                                                                             
778   RT @dog_rates: Like father (doggo), like son (pupper). Both 12/10 https://t.co/pG2inLaOda                                                                              
822   RT @dog_rates: This is just 

Checking the urls, one dog is actually only pupper, but others contains a photo of a doggo with his pupper and one isn't actually a dog, so should I convert the columns to one column dog_stage or should I convert each to boolean and clean the incorrect ones.

In [151]:
pd.melt(dogs_clean, id_vars=['tweet_id'], value_vars=['doggo', 'puppo', 'pupper'])['variable']

0       doggo 
1       doggo 
2       doggo 
3       doggo 
4       doggo 
        ...   
7063    pupper
7064    pupper
7065    pupper
7066    pupper
7067    pupper
Name: variable, Length: 7068, dtype: object